### PR TITLE
rtt: fix #1248, RTT for F072 (STM32F072) platform not working

### LINKED ISSUE #1248: RTT for F072 (STM32F072) platform not working
I have a black magic debugger with STM32F072 on a custom board. RTT logging doesn't seem to work. It works for erasing and writing just fine currently.  Tried it with both v1.8.1 and the 9f9d42861dd2772b25557fb13cbae7e56c09d20c commit in main.

I've followed the instructions in [UsingRTT.md](https://github.com/blackmagic-debug/blackmagic/blob/main/UsingRTT.md) with minicom/picocom with gdb; I can attach and enable rtt but as soon as I run it my debugger goes silent and the target will not respond to interrupt requests. The same commands for a STM32 "bluepill" board seems to work just fine.

Has anyone managed to get RTT logging to work for STM32F072? Is it a missing feature or perhaps just a problem on my end? Here's my output; it takes a long time for the F072 to work again.

```bash
GNU gdb (GNU Tools for Arm Embedded Processors 9-2019-q4-major) 8.3.0.20190709-git
Copyright (C) 2019 Free Software Foundation, Inc.
(gdb) target external-remote /dev/ttyACM0
Undefined target command: "external-remote /dev/ttyACM0".  Try "help target".
(gdb) target extended-remote /dev/ttyACM0
Remote debugging using /dev/ttyACM0
(gdb) monitor swdp_scan
Target voltage: ABSENT!
Available Targets:
No. Att Driver
 1      Nordic nRF52 M4
 2      Nordic nRF52 Access Port 
(gdb) att 1
Attaching to Remote target
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0x0003014e in ?? ()
(gdb) mon rtt enable
(gdb) run
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program:  
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
^C^CThe target is not responding to interrupt requests.
Stop debugging it? (y or n) y
Disconnected from target.
(gdb) mon rtt status
"monitor" command not supported by this target.
```

Thanks

### PR DESCRIPTION
**Pull request by koendv**
## Detailed description

Tentative fix, issue "RTT for F072 (STM32F072) platform not working #1248."

The problem seems to be two errors:

- the nrf52 target announces more ram to gdb than physically present,
  and bmp hangs when rtt reads ram that is not physically present.
  target announcing the exact amount of ram is not practical.
  solution: add an option to restrict the address range rtt searches.
  ``mon rtt ram startaddress stopaddress``
- current rtt fetches address and size of circular buffer once, at initialisation, and head and tail pointers afterwards.
  if target software modifies buffer address or buffer size after rtt initialisation, bmp has out-of-date pointers to rtt circular buffers.
  solution: before accessing rtt circular buffers on target, each time fetch full control block from target. check rtt control block present and not corrupted.

Additionally:
- cosmetic changes:
  in "mon rtt status" change `ident "off"` to `ident off`
  in "mon rtt cblock" show head and tail values, not head and tail addresses; more intuitive.
- changed variable names etc. as suggested in code review

This PR has been tested on blackpillv2; needs to be tested on F072.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

### CODE REVIEW

File: src/rtt.c [hunk 1]
```diff
@@ -411,17 +413,35 @@ void poll_rtt(target *cur_target)
...
-					else if (v == RTT_ERR) rtt_err = true;
+			/* check control block not changed or corrupted */
+			if (target_mem_read(cur_target, cblock_header, rtt_cbaddr, sizeof(cblock_header))
+				|| memcmp(saved_cblock_header, cblock_header, sizeof(cblock_header)))
+				rtt_found = false; // force searching control block next poll_rtt()
+		}
+		/* do rtt i/o if control block found */
+		if (rtt_found && rtt_cbaddr) {
+			/* copy control block from target */
+			uint32_t rtt_cblock_size = (sizeof rtt_channel[0]) * (rtt_num_up_chan + rtt_num_down_chan);
```
Comment (dragonmux) [inline src/rtt.c:425]: `sizeof(rtt_channel[0])` which also removes the outer parens around the expression

File: src/rtt.c [hunk 2]
```diff
@@ -411,17 +413,35 @@ void poll_rtt(target *cur_target)
...
+			if (target_mem_read(cur_target, rtt_channel, rtt_cbaddr + 24, rtt_cblock_size)) {
+				gdb_outf("rtt: read fail at 0x%" PRIx32 "\r\n", rtt_cbaddr + 24);
+				rtt_err = true;
+			} else {
+				for (uint32_t i = 0; i < rtt_num_up_chan + rtt_num_down_chan; i++) {
+					rtt_retval v;
+					if (rtt_channel_enabled[i]) {
+						if (i < rtt_num_up_chan) {
```
Comment (dragonmux) [inline src/rtt.c:433]: Please remove the `{}`'s

File: src/rtt.c [hunk 3]
```diff
@@ -411,17 +413,35 @@ void poll_rtt(target *cur_target)
...
+							/* rtt from target to host */
+							v = print_rtt(cur_target, i);
+						} else {
+							/* rtt from host to target */
+							rtt_flag_skip = rtt_channel[i].flag == 0;
+							rtt_flag_block = rtt_channel[i].flag == 2;
+							v = read_rtt(cur_target, i);
+						}
+						if (v == RTT_OK) rtt_busy = true;
```
Comment (dragonmux) [inline src/rtt.c:442]: Run `clang-format`. These must be written as:
```c
if (v == RTT_OK)
    rtt_busy = true;
```
etc

**Review by dragonmux — CHANGES_REQUESTED** [review]
Thank you for the contribution! This is shaping up well. We've spotted mostly stylistic things in the review with only one organisational point we'd like addressed.

**Comment by koendv** [comment]
The comment about target_mem_access_needs_halt() to the target layer is correct. But I would prefer if the riscv branch was asked their opinion, as the code is specific to that target.

**Comment by dragonmux** [comment]
We're the one working on making the RISC-V support PR-able - see branch feature/risc-v - and we're the maintainer telling you to put that function in the target layer. RISC-V is just as bound as anything else to the standards of code organisation the project has and if it's doing bad things we'll be fixing them before the final PR for v2.0.

File: src/command.c [hunk 4]
```diff
@@ -555,6 +564,15 @@ static bool cmd_rtt(target *t, int argc, const char **argv)
 			if (rtt_ident[i] == '_')
 				rtt_ident[i] = ' ';
 		}
+	} else if (argc == 2 && !strncmp(argv[1], "ram", command_len)) {
+		rtt_flag_ram = false;
+	} else if (argc == 4 && !strncmp(argv[1], "ram", command_len)) {
+		int cnt1 = sscanf(argv[2], "%" SCNx32, &rtt_ram_start);
+		int cnt2 = sscanf(argv[3], "%" SCNx32, &rtt_ram_end);
+		rtt_flag_ram = (cnt1 == 1) && (cnt2 == 1) && (rtt_ram_end > rtt_ram_start);
```
[the remark this answers is not in the document]
Comment (dragonmux) [reply src/command.c:566]: To clarify, we're saying please rewrite this to `rtt_flag_ram = cnt1 == 1 && cnt2 == 1 && rtt_ram_end > rtt_ram_start;` so the GCC warning about accidental assignment in a comparison context is enabeld

File: src/command.c [hunk 5]
```diff
@@ -499,63 +499,74 @@ static bool cmd_rtt(target *t, int argc, const char **argv)
 {
 	(void)t;
 	const size_t command_len = strlen(argv[1]);
-	if (argc == 1 || (argc == 2 && !strncmp(argv[1], "enabled", command_len))) {
+	if (argc == 1 || (argc == 2 && strncmp(argv[1], "enabled", command_len) == 0)) {
 		rtt_enabled = true;
 		rtt_found = false;
-	} else if ((argc == 2) && !strncmp(argv[1], "disabled", command_len)) {
+		memset(rtt_channel, 0, sizeof(rtt_channel));
+	} else if ((argc == 2) && strncmp(argv[1], "disabled", command_len) == 0) {
```
Comment (dragonmux) [inline src/command.c:506]: Please remove the extra parentheses from around the argc comparison here, and in the "status" block

File: src/rtt.c [hunk 6]
```diff
@@ -73,45 +81,38 @@ static char xmit_buf[RTT_UP_BUF_SIZE];
...
+	static const uint64_t p = 0x444110cd;
+	static const uint32_t stride = 128;
 	uint64_t t = 0;
-	uint8_t srch_buf[m+stride];
+	uint8_t srch_buf[m + stride];
 
-	for (struct target_ram *r = cur_target->ram; r; r = r->next) {
-		const uint32_t ram_start = r->start;
-		const uint32_t ram_end = r->start + r->length;
+	t = 0;
```
Comment (dragonmux) [inline src/rtt.c:94]: This looks like a redundant assignment as it's already taken care of by line 91?

File: src/rtt.c [hunk 7]
```diff
@@ -120,24 +121,16 @@ uint32_t memsrch(target *cur_target)
...
-				}
+	for (uint32_t addr = ram_start; addr < ram_end; addr += sizeof(srch_buf) - srch_str_len - 1) {
+		uint32_t buf_siz = MIN(ram_end - addr, sizeof(srch_buf));
+		if (target_mem_read(cur_target, srch_buf, addr, buf_siz)) {
+			gdb_outf("rtt: read fail at 0x%" PRIx32 "\r\n", addr);
+			continue;
+		}
+		for (uint32_t offset = 0; offset + srch_str_len + 1 < buf_siz; offset++) {
+			if (strncmp((const char *)(srch_buf + offset), srch_str, srch_str_len) == 0) {
+				uint32_t cb_addr = addr + offset;
```
Comment (dragonmux) [inline src/rtt.c:132]: `const` or perhaps even merge the calculation into the return statement, dropping the {}'s for the if block?

**Review by dragonmux — CHANGES_REQUESTED** [review]
This is looking much better, thank you! There are still a couple of additional items outstanding from the first review and a couple more things that re-reviewing picked up, but this looks technically good and stylistically much more consistent.

We look forward to being able to merge this soon

**Comment by koendv** [comment]
OK. That ought to do it, I think.

File: src/rtt.c [hunk 8]
```diff
@@ -236,40 +249,32 @@ static void find_rtt(target *cur_target)
...
-	buf_tail = head_tail[1];
-
-	if (buf_head >= rtt_channel[i].buf_size || buf_tail >= rtt_channel[i].buf_size)
+	if (rtt_channel[i].head >= rtt_channel[i].buf_size || rtt_channel[i].tail >= rtt_channel[i].buf_size)
 		return RTT_ERR;
 
 	/* write recv_buf to target rtt 'down' buf */
-	while ((next_head = ((buf_head + 1) % rtt_channel[i].buf_size)) != buf_tail && (ch = rtt_getchar()) != -1) {
-		if (target_mem_write(cur_target, rtt_channel[i].buf_addr + buf_head, &ch, 1))
+	while ((next_head = ((rtt_channel[i].head + 1) % rtt_channel[i].buf_size)) != rtt_channel[i].tail && (ch = rtt_getchar()) != -1) {
```
[the remark this answers is not in the document]
Comment (dragonmux) [reply src/rtt.c]: We'll note this has not been fixed, but we'll go with the change as it is and make this not block. If you figure out a way to express this while loop condition more simply with the assignments done outside the condition body, please open another PR to address it. If not, we'll give it a try on the clean-up branches we're working on. If we manage to figure something out, we'll ping you to test it and sanity check it.

**Review by dragonmux — APPROVED** [review]
LGTM! Thank you for your hard work and bearing with us. Merging

**Pull request merged**